File: 207-course-schedule/207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] indegree = new int[numCourses];
        List<List<Integer>> graph = new ArrayList<>();
        for(int i = 0;i < numCourses; i++)
            graph.add(new ArrayList<>());
        
        for(int[] pre : prerequisites){
            int u  = pre[0] , v = pre[1];
            graph.get(v).add(u);
            indegree[u]++;
        }
        int count = 0;
        Queue<Integer> q = new LinkedList<>();
        for(int i = 0; i < numCourses; i++){
             if(indegree[i] == 0) 
                q.add(i);
        }
           
        while(!q.isEmpty()){
            int curr = q.poll();
            count++;
            for(int adj : graph.get(curr)){
                if(--indegree[adj] == 0)
                    q.add(adj);
            }
        }
        
        return count == numCourses;
    }
}

/*

prereq[i] = [ai,bi] : bi -> ai

*/

File: 207-course-schedule/207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        courses_taken = 0 
        graph = defaultdict(list) 
        indegree = defaultdict(int) 
        
        for a,b in prerequisites: 
            graph[b].append(a) 
            indegree[a] += 1 
        
        q = deque()
        
        for node in range(numCourses):
            if indegree[node] == 0:
                q.append(node) 
        #print(q)
        while q:
            curr = q.popleft() 
            courses_taken += 1 
            
            for adj in graph[curr]:
                indegree[adj] -= 1 
                if indegree[adj] == 0:
                    q.append(adj) 
        
        if courses_taken == numCourses: 
            return True 
        else:
            return False 